### DISCUSSION TITLE
Subsequent calls to simpleview::render() return null

### DISCUSSION THREAD

**Discussion by quasipickle** [phalcon/cphalcon#15521] [Q&A]
Upgrading from 3.4 to 4.1.2.  I'm using simpleView to render plain text and HTML output for email.  I'm discovering, though, that only the first call to `simpleview::render()` displays anything.  Subsequent calls return NULL.  Any ideas why?

Here's my example code:
```php
echo \Phalcon\DI::getDefault()->get('simpleview')->render('path/tpl.phtml',['reason'=>$reason]);
echo \Phalcon\DI::getDefault()->get('simpleview')->render('path/tpl.phtml',['reason'=>$reason]);
```
The rendered result only gets output once.

Here is my `simpleview` service definition:

```php
$DI->set('simpleview',function() use($Config){
    $View = new \Phalcon\Mvc\View\Simple();
    $View->setViewsDir($Config->dirs->views);
    $View->registerEngines(['.phtml'=>'volt']);
    return $View;
});
```

And the `volt` service definition:
```php
$DI->set('volt',function(\Phalcon\Mvc\ViewBaseInterface $View) use($Config,$DI){
    $Volt = new \Phalcon\Mvc\View\Engine\Volt($View,$DI);
    $Volt->setOptions([ // Duplicate view directory structure when caching
                        'path' => function ($templatePath) use($View,$Config) {
                            $dirName  = dirName(substr($templatePath, strlen($View->getViewsDir())));
                            $baseName = basename($templatePath);
                            $basePath = $View->getViewsDir().'compiled/';

                            if (!is_dir($basePath.$dirName)) {
                                mkdir($basePath.$dirName, 0777, true);
                            }

                            return $basePath.$dirName.'/c_'.$baseName.'.php';
                        },
                        'always'     =>  $Config->views_always_compile,
                      ]);
    return $Volt;
});
```

**Comment by quasipickle** [comment]
To limp past the problem I did just what was suggested back on the old forums.  I moved the code that defined the Volt service to a function.  I then called that function once when defining the "Volt" service, which is used by the "View" service, and I set up my "simpleview" service to call that function each time the simpleview service is invoked.

I'm still looking for a real solution.

**Reply by wurst-hans** [reply]
Dylan, I think it has to do something with shared vs. non-shared services. When you attach your services to DI via `$di->set()` the default flag for shared is `false`, see [here](https://github.com/phalcon/cphalcon/blob/master/phalcon/Di.zep#L541). So try to add service via `setShared()` or set last parameter in `set()` to `true`.

**Reply by quasipickle** [reply]
`setShared()` means subsequent calls to the service use the same object.  `set()` means every call creates a new object.  I'm not sure why creating a new object would have any effect.

**Comment by Izopi4a** [comment]
i do it this way in phalcon 5
```
$this->simpleView = $this->getDI()
            ->get("view");

$simpleView = clone $this->simpleView; // <-- might not be needed, i used it as u with simple view in phalcon 3
        $simpleView->setRenderLevel(9);
        $simpleView->setVar("bag", $bag);

        $simpleView->render("PublicAPI", "slots")
            ->finish();
        $html= $simpleView->getContent();
```
